src/main/java/com/geoxp/geo/GeoXPShapeHelper.java

```diff
@@ -0,0 +1,261 @@
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+package com.geoxp.geo;
+
+import java.util.Arrays;
+import java.util.BitSet;
+
+import com.geoxp.GeoXPLib;
+import com.geoxp.GeoXPLib.GeoXPShape;
+
+public class GeoXPShapeHelper {
+  
+  private static long[] RESOLUTION_MASKS;
+  
+  static {
+    RESOLUTION_MASKS = new long[16];
+    RESOLUTION_MASKS[1]  = 0x0f00000000000000L;
+    RESOLUTION_MASKS[2]  = 0x0ff0000000000000L;
+    RESOLUTION_MASKS[3]  = 0x0fff000000000000L;
+    RESOLUTION_MASKS[4]  = 0x0ffff00000000000L;
+    RESOLUTION_MASKS[5]  = 0x0fffff0000000000L;
+    RESOLUTION_MASKS[6]  = 0x0ffffff000000000L;
+    RESOLUTION_MASKS[7]  = 0x0fffffff00000000L;
+    RESOLUTION_MASKS[8]  = 0x0ffffffff0000000L;
+    RESOLUTION_MASKS[9]  = 0x0fffffffff000000L;
+    RESOLUTION_MASKS[10] = 0x0ffffffffff00000L;
+    RESOLUTION_MASKS[11] = 0x0fffffffffff0000L;
+    RESOLUTION_MASKS[12] = 0x0ffffffffffff000L;
+    RESOLUTION_MASKS[13] = 0x0fffffffffffff00L;
+    RESOLUTION_MASKS[14] = 0x0ffffffffffffff0L;
+    RESOLUTION_MASKS[15] = 0x0fffffffffffffffL;
+  }
+
+  public static boolean intersects(GeoXPShape a, GeoXPShape b) {
+    return null != intersection(a, b, true);
+  }
+  
+  public static GeoXPShape intersection(GeoXPShape a, GeoXPShape b) {
+    return intersection(a, b, false);
+  }
+  
+  private static GeoXPShape intersection(GeoXPShape a, GeoXPShape b, boolean check) {
+       
+   long[] acells = GeoXPLib.getCells(a);
+   long[] bcells = GeoXPLib.getCells(b);
+   
+   //
+   // Sort cells
+   //
+   //Arrays.sort(acells);
+   //Arrays.sort(bcells);
+   
+   // Determine the start/end index of each resolution
+   int[] astart = new int[16];
+   int[] aend = new int[16];
+   int[] bstart = new int[16];
+   int[] bend = new int[16];
+   
+   for (int i = 1; i < 16; i++) {
+     long cell = ((long) i) << 60;
+     int idx = Arrays.binarySearch(acells, cell); 
+
+     if (idx >= 0) {
+       astart[i] = idx;
+     } else {
+       astart[i] = -idx - 1;
+       if (astart[i] >= acells.length || (acells[astart[i]] & 0xF000000000000000L) != cell) {
+         astart[i] = -1;
+       }
+     }
+            
+     idx = Arrays.binarySearch(bcells, cell);       
+     
+     if (idx >= 0) {
+       bstart[i] = idx;
+     } else {
+       bstart[i] = -idx - 1;
+       if (bstart[i] >= bcells.length || (bcells[bstart[i]] & 0xF000000000000000L) != cell) {
+         bstart[i] = -1;
+       }
+     }
+   }
+   
+   for (int i = 1; i < 16; i++) {
+     aend[i] = -1;
+     if (-1 != astart[i]) {
+       // Find the start of the next index
+       for (int j = i + 1; j < 16; j++) {
+         if (-1 != astart[j]) {
+           aend[i] = astart[j] - 1;
+           break;
+         }
```

Comment:
This won't work because `aend[i] = astart[j] - 1` does not hold for `i < 8` and `j >= 8`.